Document known testcontainers lifecycle issues

See https://github.com/spring-projects/spring-boot/issues/35223#issuecomment-1531061583

Hi, in the current state I don't know if what I'm doing is wrong, but I've tried bean approach and static field approach with static init block. Tests are passing connecting to mongo, but the build produces multiple errors, that it cannot connect to mongo. Maybe after close of the container. I have multi module spring boot project. I've used bean approach described here, but it does not seem to work fully as it is posting those multiple build errors about prematurely closing connection and such: https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#features.testcontainers.at-development-time
My current approach is:


```
import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.testcontainers.containers.MongoDBContainer;
import org.testcontainers.containers.Network;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.utility.MountableFile;

import java.io.IOException;

@Slf4j
@SpringBootTest
public class SpringBootTestBase {

    static final MongoDBContainer mongoDbContainer = mongoDbContainer();
    static final PostgreSQLContainer<?> postgresDbContainer = postgresDbContainer();

    private static final String DEFAULT_IMAGE_AND_TAG = "mongo:6.0.1";
    private static final String RESOURCE_NAME = "rs-init.sh";
    private static final String CONTAINER_SCRIPT_PATH = "/scripts/rs-init.sh";

    static MongoDBContainer mongoDbContainer() {
        int mongoDbPort = 27017;
        var network = Network.builder().driver("bridge").build();

        return new MongoDBContainer(DEFAULT_IMAGE_AND_TAG)
                .withCommand("/usr/bin/mongod", "--bind_ip_all", "--replSet", "dbrs")
                .withNetwork(network)
                .withExposedPorts(mongoDbPort)
                .withCopyFileToContainer(MountableFile.forClasspathResource(RESOURCE_NAME), CONTAINER_SCRIPT_PATH);

    }

    static PostgreSQLContainer<?> postgresDbContainer() {
        return new PostgreSQLContainer<>("postgres:15.2")
                .withDatabaseName("test")
                .withUsername("postgres")
                .withPassword("abcd");
    }

    @DynamicPropertySource
    static void dynamicProperties(DynamicPropertyRegistry registry) {
        log.info("=== MONGO DB STARTING ===");
        mongoDbContainer.start();
        try {
            mongoDbContainer.execInContainer(CONTAINER_SCRIPT_PATH);
        } catch (IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }
        registry.add("spring.data.mongodb.host", mongoDbContainer::getHost);
        registry.add("spring.data.mongodb.port", mongoDbContainer::getFirstMappedPort);

        log.info("===  POSTGRES STARTING ===");
        postgresDbContainer.start();
        registry.add("spring.datasource.url", postgresDbContainer::getJdbcUrl);
        registry.add("spring.datasource.username", postgresDbContainer::getUsername);
        registry.add("spring.datasource.password", postgresDbContainer::getPassword);
    }

}
```
Do note, that only thing I'm doing is extending this class in the test files not annotated with `@SpringBootTest` and that is it. Maybe I should do something more here?
and rs init file is:

```
#!/bin/bash

mongosh <<EOF
var config = {
      "_id": "dbrs",
      "version": 1,
      "members": [
          {
              "_id": 1,
              "host": "localhost:27017"
          }
      ]
    };
    rs.initiate(config, { force: true });
EOF

sleep 2

mongosh <<EOF
    use admin;
    someDbName = db.getSiblingDB("someDbName");
    someDbName.createUser(
        {
            user: "someUser",
            pwd: "somePassword",
            roles: [ { role: "readWrite", db: "someDbName" } ],
            passwordDigestor: "server"
    });

    rs.status();
EOF
```
Could you please help me with this issue? It all happend after upgrade to spring boot 3.2.1. I get multiple issues in logs like this one below:

```
com.mongodb.MongoSocketReadException: Exception receiving message
	at com.mongodb.internal.connection.InternalStreamConnection.translateReadException(InternalStreamConnection.java:730)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveMessageWithAdditionalTimeout(InternalStreamConnection.java:606)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:451)
	at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:372)
	at com.mongodb.internal.connection.CommandHelper.sendAndReceive(CommandHelper.java:102)
	at com.mongodb.internal.connection.CommandHelper.executeCommand(CommandHelper.java:49)
	at com.mongodb.internal.connection.InternalStreamConnectionInitializer.initializeConnectionDescription(InternalStreamConnectionInitializer.java:130)
	at com.mongodb.internal.connection.InternalStreamConnectionInitializer.startHandshake(InternalStreamConnectionInitializer.java:78)
	at com.mongodb.internal.connection.InternalStreamConnection.open(InternalStreamConnection.java:213)
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.lookupServerDescription(DefaultServerMonitor.java:196)
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.run(DefaultServerMonitor.java:156)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.net.SocketException: Connection reset
	at java.base/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:328)
	at java.base/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:355)
	at java.base/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:808)
	at java.base/java.net.Socket$SocketInputStream.read(Socket.java:966)
	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:175)
	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:200)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveResponseBuffers(InternalStreamConnection.java:739)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveMessageWithAdditionalTimeout(InternalStreamConnection.java:603)
```
The strange thing is it happens also in testClasses so compilation, this is very strange in this case. But it happens also after tests. Example below:

```
> Task :module-name:testClasses
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
2024-01-10 12:14:28.727  INFO 112868 --- [localhost:32769] org.mongodb.driver.cluster               : Exception in monitor thread while connecting to server localhost:32769

com.mongodb.MongoSocketReadException: Prematurely reached end of stream
	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:177)
	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:200)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveResponseBuffers(InternalStreamConnection.java:739)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveMessageWithAdditionalTimeout(InternalStreamConnection.java:603)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:451)
	at com.mongodb.internal.connection.InternalStreamConnection.receive(InternalStreamConnection.java:404)
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.lookupServerDescription(DefaultServerMonitor.java:224)
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.run(DefaultServerMonitor.java:156)
	at java.base/java.lang.Thread.run(Thread.java:840)

2024-01-10 12:14:28.729  INFO 112868 --- [localhost:32769] org.mongodb.driver.cluster               : Exception in monitor thread while connecting to server localhost:32769

com.mongodb.MongoSocketReadException: Exception receiving message
	at com.mongodb.internal.connection.InternalStreamConnection.translateReadException(InternalStreamConnection.java:730)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveMessageWithAdditionalTimeout(InternalStreamConnection.java:606)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:451)
	at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:372)
	at com.mongodb.internal.connection.CommandHelper.sendAndReceive(CommandHelper.java:102)
	at com.mongodb.internal.connection.CommandHelper.executeCommand(CommandHelper.java:49)
	at com.mongodb.internal.connection.InternalStreamConnectionInitializer.initializeConnectionDescription(InternalStreamConnectionInitializer.java:130)
	at com.mongodb.internal.connection.InternalStreamConnectionInitializer.startHandshake(InternalStreamConnectionInitializer.java:78)
	at com.mongodb.internal.connection.InternalStreamConnection.open(InternalStreamConnection.java:213)
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.lookupServerDescription(DefaultServerMonitor.java:196)
	at com.mongodb.internal.connection.DefaultServerMonitor$ServerMonitorRunnable.run(DefaultServerMonitor.java:156)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.net.SocketException: Connection reset
	at java.base/sun.nio.ch.NioSocketImpl.implRead(NioSocketImpl.java:328)
	at java.base/sun.nio.ch.NioSocketImpl.read(NioSocketImpl.java:355)
	at java.base/sun.nio.ch.NioSocketImpl$1.read(NioSocketImpl.java:808)
	at java.base/java.net.Socket$SocketInputStream.read(Socket.java:966)
	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:175)
	at com.mongodb.internal.connection.SocketStream.read(SocketStream.java:200)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveResponseBuffers(InternalStreamConnection.java:739)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveMessageWithAdditionalTimeout(InternalStreamConnection.java:603)
	... 10 common frames omitted

> Task :module-name:test
```

Please, help if you can. @philwebb or someone.

--- a/spring-boot-project/spring-boot-docs/src/docs/antora/modules/reference/pages/testing/testcontainers.adoc
+++ b/spring-boot-project/spring-boot-docs/src/docs/antora/modules/reference/pages/testing/testcontainers.adoc
@@ -40,6 +40,11 @@ include-code::MyIntegrationTests[]
 The example above will start up a Neo4j container before any of the tests are run.
 The lifecycle of the container instance is managed by Testcontainers, as described in {url-testcontainers-docs}/test_framework_integration/junit_5/#extension[their official documentation].
 
+When using the JUnit extension, container instances are stopped after the test class has run (for static fields) or after each test method (for non-static fields).
+This can cause issues when used with Spring Boot tests, as Spring's TestContext Framework may cache the javadoc:org.springframework.context.ApplicationContext[] beyond that point and reuse it for another test class or method with the same configuration.
+If the cached application context contains beans that depend on a container that has already been stopped, later tests or bean destruction callbacks may fail.
+For this reason, you should prefer xref:#testing.testcontainers.spring-beans[managing containers as Spring beans] or xref:#testing.testcontainers.importing-configuration-interfaces[importing container declarations] when the application context should remain usable for as long as it is cached.
+
 NOTE: In most cases, you will additionally need to configure the application to connect to the service running in the container.
 
 
@@ -82,19 +87,15 @@ TIP: When your application beans rely on functionality of containers, prefer con
 NOTE: Having containers managed by Testcontainers instead of as Spring beans provides no guarantee of the order in which beans and containers will shutdown.
 It can happen that containers are shutdown before the beans relying on container functionality are cleaned up.
 This can lead to exceptions being thrown by client beans, for example, due to loss of connection.
-
-WARNING: When using Testcontainers' JUnit extension with static fields annotated with javadoc:org.testcontainers.junit.jupiter.Container[format=annotation], container instances are stopped after the test class has run.
-Spring's TestContext Framework may cache the javadoc:org.springframework.context.ApplicationContext[] beyond that point and reuse it for another test class with the same configuration.
-If the cached application context contains beans that depend on a container that has already been stopped, later tests or bean destruction callbacks may fail.
 When the application context should remain usable for as long as it is cached, prefer managing containers as Spring beans or importing container declarations with javadoc:org.springframework.boot.testcontainers.context.ImportTestcontainers[format=annotation].
 
 Container beans are created and started once per application context managed by Spring's TestContext Framework.
-For details about how TestContext Framework manages the underlying application contexts and beans therein, please refer to the {url-spring-framework-docs}[Spring Framework documentation].
+For details about how the TestContext Framework manages the underlying application contexts and beans therein, please refer to the {url-spring-framework-docs}[Spring Framework documentation].
 
 Container beans are stopped as part of the TestContext Framework's standard application context shutdown process.
 When the application context gets shutdown, the containers are shutdown as well.
 This usually happens after all tests using that specific cached application context have finished executing.
-It may also happen earlier, depending on the caching behavior configured in TestContext Framework.
+It may also happen earlier, depending on the caching behavior configured in the TestContext Framework.
 
 NOTE: A single test container instance can, and often is, retained across execution of tests from multiple test classes.
 

--- a/spring-boot-project/spring-boot-docs/src/docs/antora/modules/reference/pages/testing/testcontainers.adoc
+++ b/spring-boot-project/spring-boot-docs/src/docs/antora/modules/reference/pages/testing/testcontainers.adoc
@@ -83,6 +83,11 @@ NOTE: Having containers managed by Testcontainers instead of as Spring beans pro
 It can happen that containers are shutdown before the beans relying on container functionality are cleaned up.
 This can lead to exceptions being thrown by client beans, for example, due to loss of connection.
 
+WARNING: When using Testcontainers' JUnit extension with static fields annotated with javadoc:org.testcontainers.junit.jupiter.Container[format=annotation], container instances are stopped after the test class has run.
+Spring's TestContext Framework may cache the javadoc:org.springframework.context.ApplicationContext[] beyond that point and reuse it for another test class with the same configuration.
+If the cached application context contains beans that depend on a container that has already been stopped, later tests or bean destruction callbacks may fail.
+When the application context should remain usable for as long as it is cached, prefer managing containers as Spring beans or importing container declarations with javadoc:org.springframework.boot.testcontainers.context.ImportTestcontainers[format=annotation].
+
 Container beans are created and started once per application context managed by Spring's TestContext Framework.
 For details about how TestContext Framework manages the underlying application contexts and beans therein, please refer to the {url-spring-framework-docs}[Spring Framework documentation].
 

